frame/glutton/src/lib.rs

```diff
@@ -0,0 +1,259 @@
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event {
+		/// The pallet has been initialized by root.
+		PalletInitialized,
+		/// The computation limit has been updated by root.
+		ComputationLimitSet { compute: Perbill },
+		/// The storage limit has been updated by root.
+		StorageLimitSet { storage: Perbill },
+	}
+
+	#[pallet::storage]
+	pub(crate) type Compute<T: Config> = StorageValue<_, Perbill, ValueQuery>;
+
+	#[pallet::storage]
+	pub(crate) type Storage<T: Config> = StorageValue<_, Perbill, ValueQuery>;
+
+	/// Storage map used for wasting proof size.
+	///
+	/// It contains no meaningful data - hence the name "Trash". The maximal number of entries is
+	/// set to 65k, which is just below the next jump at 16^4. This is important to reduce the proof
+	/// size benchmarking overestimate. The assumption here is that we won't have more than 65k *
+	/// 1KiB = 65MiB of proof size wasting in practice. However, this limit is not enforces, so the
+	/// pallet would also work out of the box with more entries, but its benchmarked proof weight
+	/// would possibly be underestimated in that case.
+	#[pallet::storage]
+	pub(super) type TrashData<T: Config> = StorageMap<
+		Hasher = Twox64Concat,
+		Key = u32,
+		Value = [u8; 1024],
+		QueryKind = OptionQuery,
+		MaxValues = ConstU32<65_000>,
+	>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn integrity_test() {
+			assert!(
+				!T::WeightInfo::waste_ref_time_iter(1).ref_time().is_zero(),
+				"Weight zero; would get stuck in an infinite loop"
+			);
+			assert!(
+				!T::WeightInfo::waste_proof_size_some(1).proof_size().is_zero(),
+				"Weight zero; would get stuck in an infinite loop"
+			);
+		}
+
+		fn on_idle(_: BlockNumberFor<T>, remaining_weight: Weight) -> Weight {
+			let mut meter = WeightMeter::from_limit(remaining_weight);
+			if !meter.check_accrue(T::WeightInfo::empty_on_idle()) {
+				return T::WeightInfo::empty_on_idle()
+			}
+
+			let proof_size_limit = Storage::<T>::get().mul_floor(meter.remaining().proof_size());
+			let computation_weight_limit =
+				Compute::<T>::get().mul_floor(meter.remaining().ref_time());
+			let mut meter = WeightMeter::from_limit(Weight::from_parts(
+				computation_weight_limit,
+				proof_size_limit,
+			));
+
+			Self::waste_at_most_proof_size(&mut meter);
+			Self::waste_at_most_ref_time(&mut meter);
+
+			meter.consumed
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Initializes the pallet by writing into `TrashData`.
+		///
+		/// Only callable by Root. A good default for `trash_count` is `5_000`.
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::DbWeight::get().writes((*trash_count).into()))]
+		pub fn initialize_pallet(origin: OriginFor<T>, trash_count: u32) -> DispatchResult {
+			ensure_root(origin)?;
+
+			// Fill up the `TrashData` storage item.
+			(0..trash_count).for_each(|i| TrashData::<T>::insert(i, &[i as u8; 1024]));
```

Comment:
Would it be useful to have a way to expand/shrink TrashData later?

Reply:
I think it would make sense to have a way to expand since we can have up to 65k entries inside the `StorageMap`.
Shrinking could probably be useful to free up some space after we are done testing the limits.

Reply:
Parachains can currently not use more than 5MiB proof size, but you can still add extension functionality :man_shrugging: